smart-contracts/advanced/deterministic-deployment.md

```diff
@@ -0,0 +1,141 @@
+
+- Address of the EOA;
+- Nonce of the EOA.
+
+If deploying using a `CREATE` factory contract, contract addresses are calculated using the:
+
+- Address of the factory contract (that calls `CREATE` opcode) itself;
+- Nonce of the factory contract ([EIP-161](https://eips.ethereum.org/EIPS/eip-161) specifies that contract nonce starts at 1, not 0 like EOAs).
+
+### **`CREATE2`**
+
+The `CREATE2` opcode must be run from a deployed contract, so usually it's done from a `CREATE2` factory contract. Contract addresses are precomputed using the:
+
+- Address of the factory contract (that calls `CREATE2` opcode) itself;
+- Bytecode of the contract;
+- Salt (a chosen value).
+
+See the following code snippet for an example of how to call `CREATE2` from a factory contract:
+
+```solidity
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.0;
+
+...
+
+function deploy(bytes memory bytecode, uint256 _salt) public payable {
+    address addr;
+
+    /*
+    NOTE: How to call create2
+
+    create2(v, p, n, s)
+    create new contract with code at memory p to p + n
+    and send v wei
+    and return the new address
+    where new address = first 20 bytes of keccak256(0xff + address(this) + s + keccak256(mem[p…(p+n)))
+            s = big-endian 256-bit value
+    */
+    assembly {
+        addr :=
+            create2(
+                callvalue(), // wei sent with current call
+                // Actual code starts after skipping the first 32 bytes
+                add(bytecode, 0x20),
+                mload(bytecode), // Load the size of code contained in the first 32 bytes
+                _salt // Salt from function arguments
+            )
+
+        if iszero(extcodesize(addr)) { revert(0, 0) }
+    }
+
+    emit Deployed(addr, _salt);
+}
+```
+
+### **`CREATE3`**
+
+With `CREATE2`, the contract bytecode affects the deployment address. So even blank spaces and comment text can affect the address.
+
+CREATE3 is similar to CREATE2 but without including the contract initCode on the address derivation formula. It can be used to generate deterministic contract addresses that aren’t tied to a specific contract code.
+
+CREATE3 is a way to use CREATE and CREATE2 in combination such that bytecode no longer affects the deployment address. — CREATE3 is more expensive than CREATE or CREATE2 (Fixed extra cost of ~55k gas).
+
+Check out a reference implementation [here](https://github.com/0xsequence/create3).
+
+### **Usage**
+
+Other people may have already deployed the factory contract onto some of your desired blockchains to the expected address (if they didn't change the deployment transaction data), in which case you won't need to deploy it on those blockchains - you can then just use those already-deployed factory contracts to deploy whatever other contracts you want to deploy. So first check the expected address on a blockchain explorer to see if a factory contract already exists there.
```

Comment:
"may have already deployed the factory contract to Filecoin, to the expected address" since this is specific to Filecoin and not a general article. Subsequent edits will be needed on the paragraph to match the above edit.